nodeInterface/NodeInterface.go

```diff
@@ -590,3 +590,83 @@ func (n NodeInterface) LegacyLookupMessageBatchProof(c ctx, evm mech, batchNum h
 	calldataForL1 = data
 	return
 }
+
+func (n NodeInterface) blockL1Num(l2BlockNum uint64) (uint64, error) {
+	blockHeader, err := n.backend.HeaderByNumber(n.context, rpc.BlockNumber(l2BlockNum))
+	if err != nil {
+		return 0, err
+	}
+	blockL1Num := types.DeserializeHeaderExtraInformation(blockHeader).L1BlockNumber
+	return blockL1Num, nil
+}
+
+func (n NodeInterface) matchL2BlockNumWithL1(l2BlockNum uint64, l1BlockNum uint64) error {
+	blockL1Num, err := n.blockL1Num(l2BlockNum)
+	if err != nil {
+		return fmt.Errorf("failed to get the L1 block number of the L2 block: %v. Error: %w", l2BlockNum, err)
+	}
+	if blockL1Num != l1BlockNum {
+		return fmt.Errorf("no L2 block was found with the given L1 block number. Found L2 block: %v with L1 block number: %v, given L1 block number: %v", l2BlockNum, blockL1Num, l1BlockNum)
+	}
+	return nil
+}
+
+// L2BlockRangeForL1 finds the first and last L2 block numbers that have the given L1 block number
+func (n NodeInterface) L2BlockRangeForL1(c ctx, evm mech, l1BlockNum uint64) (uint64, uint64, error) {
+	currentBlockNum := n.backend.CurrentBlock().Number.Uint64()
+	genesis := n.backend.ChainConfig().ArbitrumChainParams.GenesisBlockNum
+
+	type helperStruct struct {
+		low  uint64
+		high uint64
+	}
+
+	searchHelper := func(currentBlock *helperStruct, fetchedMid *helperStruct, target uint64) error {
+		if currentBlock.low < currentBlock.high {
+			// dont fetch midBlockL1Num if its already fetched above
+			mid := arbmath.SaturatingUAdd(currentBlock.low, currentBlock.high) / 2
+			var midBlockL1Num uint64
+			var err error
+			if mid == fetchedMid.low {
+				midBlockL1Num = fetchedMid.high
+			} else {
+				midBlockL1Num, err = n.blockL1Num(mid)
+				if err != nil {
+					return err
+				}
+				fetchedMid.low = mid
+				fetchedMid.high = midBlockL1Num
+			}
+			if midBlockL1Num < target {
+				currentBlock.low = mid + 1
+			} else {
+				currentBlock.high = mid
+			}
+			return nil
+		}
+		return nil
+	}
+	firstBlock := &helperStruct{low: genesis, high: currentBlockNum}
+	lastBlock := &helperStruct{low: genesis, high: currentBlockNum}
+	// in storedMid low corresponds to value mid and high corresponds to midBlockL1Num inside searchHelper
+	storedMid := &helperStruct{low: currentBlockNum + 1}
+	var err error
+	for firstBlock.low < firstBlock.high || lastBlock.low < lastBlock.high {
+		if err = searchHelper(firstBlock, storedMid, l1BlockNum); err != nil {
+			return 0, 0, err
+		}
+		if err = searchHelper(lastBlock, storedMid, l1BlockNum+1); err != nil {
+			return 0, 0, err
+		}
+	}
+	if err := n.matchL2BlockNumWithL1(firstBlock.high, l1BlockNum); err != nil {
+		return 0, 0, err
+	}
+	if err := n.matchL2BlockNumWithL1(lastBlock.high, l1BlockNum); err != nil {
+		lastBlock.high -= 1
+		if err = n.matchL2BlockNumWithL1(lastBlock.high, l1BlockNum); err != nil {
+			return 0, 0, err
+		}
+	}
```

Comment:
This is still a bit more complex than it needs to be. You can simplify with something like this:

```lang=go
cache := make(map[uint64]uint64)
binarySearch := func(target uint64) (uint64, error) {
	l, r := genesis, currentBlockNum
	// check ends separately, handle err.
	while l < r + 1 {
		mid := arbmath.SaturatingUAdd(l, r) / 2
		if _, found := cache[mid]; ! found {
			v, err := n.blockL1Num(mid)
			if err != nil {
				return 0, err
			}
			cache[mid] = v
		}
		if cache[mid] < target {
			l = mid
		} else {
			r = mid
		}
	}
	return l, nil
}

l, err := binarySearch(l1BlockNum)
// handle error
r, err := binarySearch(l1BlockNum+1)
// handle error
return l, r, nil
```


That is, you don't need to pass that helperStruct and have entire binary search logic in the function rather than some part in anonymous function and some part outside in the for loop.

Reply:
The time complexity has always been the same log(n) with same constant. The concern (from Lee) was number of queries to look up header by number and this approach does at most the same amount of queries as your initial approach.

That is correct, this way we use cache, which is completely negligible (both memory and runtime-wise), but code is more readable. When runtime/space overhead is negligible we should aim for more readable code.

Reply:
Cool